packages/techdocs-common/src/stages/publish/openStackSwift.ts

```diff
@@ -0,0 +1,263 @@
+        const uploadFile = limiter(
+          () =>
+            new Promise((res, rej) => {
+              const readStream = fs.createReadStream(filePath, 'utf8');
+
+              const writeStream = this.storageClient.upload(params);
+
+              writeStream.on('error', rej);
+
+              writeStream.on('success', res);
+
+              readStream.pipe(writeStream);
+            }),
+        );
+        uploadPromises.push(uploadFile);
+      }
+      await Promise.all(uploadPromises);
+      this.logger.info(
+        `Successfully uploaded all the generated files for Entity ${entity.metadata.name}. Total number of files: ${allFilesToUpload.length}`,
+      );
+      return;
+    } catch (e) {
+      const errorMessage = `Unable to upload file(s) to OpenStack Swift. ${e}`;
+      this.logger.error(errorMessage);
+      throw new Error(errorMessage);
+    }
+  }
+
+  async fetchTechDocsMetadata(
+    entityName: EntityName,
+  ): Promise<TechDocsMetadata> {
+    try {
+      return await new Promise<TechDocsMetadata>(async (resolve, reject) => {
+        const entityRootDir = `${entityName.namespace}/${entityName.kind}/${entityName.name}`;
+
+        const stream = this.storageClient.download({
+          container: this.containerName,
+          remote: `${entityRootDir}/techdocs_metadata.json`,
+        });
+
+        try {
+          const techdocsMetadataJson = await streamToBuffer(stream);
+          if (!techdocsMetadataJson) {
+            throw new Error(
+              `Unable to parse the techdocs metadata file ${entityRootDir}/techdocs_metadata.json.`,
+            );
+          }
+
+          const techdocsMetadata = JSON5.parse(
+            techdocsMetadataJson.toString('utf-8'),
+          );
+
+          resolve(techdocsMetadata);
+        } catch (err) {
+          this.logger.error(err.message);
+          reject(new Error(err.message));
+        }
+      });
+    } catch (e) {
+      throw new Error(`TechDocs metadata fetch failed, ${e.message}`);
+    }
+  }
+
+  /**
+   * Express route middleware to serve static files on a route in techdocs-backend.
+   */
+  docsRouter(): express.Handler {
+    return async (req, res) => {
+      // Trim the leading forward slash
+      // filePath example - /default/Component/documented-component/index.html
+
+      const filePath = req.path.replace(/^\//, '');
+
+      // Files with different extensions (CSS, HTML) need to be served with different headers
+      const fileExtension = path.extname(filePath);
+      const responseHeaders = getHeadersForFileExtension(fileExtension);
+
+      const stream = this.storageClient.download({
+        container: this.containerName,
+        remote: filePath,
+      });
+
+      try {
+        // Inject response headers
+        for (const [headerKey, headerValue] of Object.entries(
+          responseHeaders,
+        )) {
+          res.setHeader(headerKey, headerValue);
+        }
+
+        res.send(await streamToBuffer(stream));
```

Comment:
Kinda minor, but is there any reason not to pipe `this.storageClient.download()` directly through to `res`?  We're still buffering the whole file into memory before sending it to the client and it'd be nice to avoid the memory overhead.

The [GCS Publisher](https://github.com/backstage/backstage/blob/master/packages/techdocs-common/src/stages/publish/googleStorage.ts#L185) would be a good reference.

Reply:
We were inspired existing aws S3 publisher they do same thing. Next version we thing plan change OpenStack Swift package and some configuration. We can do this on the next version if it is no problem?